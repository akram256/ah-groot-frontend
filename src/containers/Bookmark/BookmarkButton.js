import React, { Component } from 'react';
import { connect } from 'react-redux';

import {
  bookmarkArticle,
  unBookmarkArticle,
  getAllbookmarkedArticles,
} from '../../actions/bookmarks/BookmarkAction';

export class BookmarkButton extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isBookMarked: 'unbookmarked',
    };
  }

  componentWillMount() {
    this.props.getAllbookmarkedArticles();
  }

  componentWillReceiveProps(nextProps){
    const { bookmarks } = nextProps;
    const isBookmarkedAlready = bookmarks.find((bookmark)=>(bookmark.slug === nextProps.slug));
    if(isBookmarkedAlready){
        this.setState({ isBookMarked: 'bookmarked' });
    }
  }

  render() {
    return (
      <a
        className={`large material-icons ${this.state.isBookMarked}`}
        onClick={() => {
          if (this.state.isBookMarked === 'unbookmarked') {
              this.setState({ isBookMarked: 'bookmarked' });
            this.props.bookmarkArticle(this.props.slug);
          } else {
              this.setState({ isBookMarked: 'unbookmarked' });
            this.props.unBookmarkArticle(this.props.slug);
          }
        }}
      >
        bookmark
      </a>
    );
  }
}

export const mapStateToProps = state => {
   /* istanbul ignore next */
  return {
    ...state,
    bookmark: state.bookmark,
    unbookmark: state.unbookmark,
    bookmarks: state.bookmarks,
  };
};

export const mapDispatchToProps = dispatch => {
  return {
    bookmarkArticle: function(slug) {
      dispatch(bookmarkArticle(slug));
    },
    unBookmarkArticle: function(slug) {
      dispatch(unBookmarkArticle(slug));
    },
    getAllbookmarkedArticles: function() {
      dispatch(getAllbookmarkedArticles());
    },
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(BookmarkButton);
