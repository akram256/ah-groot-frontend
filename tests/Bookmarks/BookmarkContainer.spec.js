import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow, mount } from 'enzyme';
import thunk from 'redux-thunk';
import moxios from 'moxios';

import BookmarkButton, {
  BookmarkButton as DumpBookmarkButton,
  mapDispatchToProps,
} from '../../src/containers/Bookmark/BookmarkButton';
import {
  AllBookmarks,
  mapDispatchToProps as customMapDispatchToProps,
  mapStateToProps as customMapStateToProps
 } from '../../src/containers/Bookmark/AllBookmarks';

describe('Bookmarks botton ', () => {
  const mockStore = configureStore([thunk]);
  let store;
  let bookmarkResponse;

  beforeEach(() => {
    store = mockStore([]);
    bookmarkResponse = {
      bookmark: {
        author: 'stanley',
        article_title: 'Groot is such an amazing team',
        slug: 'groot-is-such-an-amazing',
        description: 'Groot is such an amazing team',
        bookmarked_at: '2019-04-24T04:27:05.188296+03:00',
        image: '',
      },
    };
    moxios.install();
  });

  afterEach(function() {
    moxios.uninstall();
  });

  it('should render without crashing', () => {
    const wrapper = shallow(
      <DumpBookmarkButton getAllbookmarkedArticles={() => jest.fn()} />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should receive new props', () => {
    const wrapper = shallow(
      <DumpBookmarkButton
      getAllbookmarkedArticles={() => jest.fn()}
        bookmarks ={[bookmarkResponse.bookmark]}
      />
    );
    wrapper
      .instance()
      .componentWillReceiveProps({ bookmarks: [bookmarkResponse.bookmark], slug: 'groot-is-such-an-amazing' });
    expect(wrapper).toMatchSnapshot();
  });

  it('should respond to on click events', () => {
    const wrapper = shallow(
        <DumpBookmarkButton
        getAllbookmarkedArticles={() => jest.fn()}
        bookmarkArticle={() => jest.fn()}
        unBookmarkArticle={() => jest.fn()}
        bookmarks ={[bookmarkResponse.bookmark]}
        />
      );
      wrapper.find('.material-icons').simulate('click');
      wrapper.find('.material-icons').simulate('click');
  });

  it('should match state to props', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).getAllbookmarkedArticles();
    mapDispatchToProps(dispatch).unBookmarkArticle();
    mapDispatchToProps(dispatch).bookmarkArticle();
  });
});

describe('All bookmarks ', () => {
    const mockStore = configureStore([thunk]);
    let store;
    let bookmarkResponse;

    beforeEach(() => {
      store = mockStore([]);
      bookmarkResponse = {
        bookmark: {
          author: 'stanley',
          article_title: 'Groot is such an amazing team',
          slug: 'groot-is-such-an-amazing',
          description: 'Groot is such an amazing team',
          bookmarked_at: '2019-04-24T04:27:05.188296+03:00',
          image: '',
        },
      };
      moxios.install();
    });

    afterEach(function() {
      moxios.uninstall();
    });

    it('should render without crashing', () => {
      const wrapper = shallow(
        <AllBookmarks getAllbookmarkedArticles={() => jest.fn()} />
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('should receive new props', () => {
      const wrapper = shallow(
        <AllBookmarks
        getAllbookmarkedArticles={() => jest.fn()}
          bookmarks ={[bookmarkResponse.bookmark]}
        />
      );
      wrapper
        .instance()
        .componentWillReceiveProps({ bookmarks: [bookmarkResponse.bookmark] });
      expect(wrapper).toMatchSnapshot();
    });

    it('should map dispatch to props', () => {
      const dispatch = jest.fn();
      customMapDispatchToProps(dispatch).getAllbookmarkedArticles();
    });

    it('should match state to props', () => {
      const initialState = {
        bookmarks: []
      };
      customMapStateToProps(initialState);
    });

  });
