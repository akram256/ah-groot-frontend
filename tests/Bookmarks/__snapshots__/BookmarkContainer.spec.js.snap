// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All bookmarks  should receive new props 1`] = `ShallowWrapper {}`;

exports[`All bookmarks  should render without crashing 1`] = `ShallowWrapper {}`;

exports[`Bookmarks botton  should receive new props 1`] = `ShallowWrapper {}`;

exports[`Bookmarks botton  should render without crashing 1`] = `ShallowWrapper {}`;
